sql/logic_test.go

```diff
@@ -0,0 +1,321 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Peter Mattis (peter@cockroachlabs.com)
+
+package sql_test
+
+import (
+	"bufio"
+	"bytes"
+	"crypto/md5"
+	"database/sql"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"reflect"
+	"regexp"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/cockroachdb/cockroach/server"
+	"github.com/cockroachdb/cockroach/testutils"
+	"github.com/cockroachdb/cockroach/util/leaktest"
+)
+
+var (
+	resultsRE = regexp.MustCompile(`^(\d+)\s+values?\s+hashing\s+to\s+([0-9A-Fa-f]+)$`)
+	errorRE   = regexp.MustCompile(`^(?:statement|query)\s+error\s+(.*)$`)
+)
+
+type lineScanner struct {
+	*bufio.Scanner
+	line int
+}
+
+func newLineScanner(r io.Reader) *lineScanner {
+	return &lineScanner{
+		Scanner: bufio.NewScanner(r),
+		line:    0,
+	}
+}
+
+func (l *lineScanner) Scan() bool {
+	ok := l.Scanner.Scan()
+	if ok {
+		l.line++
+	}
+	return ok
+}
+
+type logicStatement struct {
+	pos       string
+	sql       string
+	expectErr string
+}
+
+type logicQuery struct {
+	pos             string
+	sql             string
+	expectErr       string
+	expectedValues  int
+	expectedHash    string
+	expectedResults []string
+}
+
+// TODO(pmattis): #1961 is adding a similar type to cli/sql.go. Perhaps move
+// this type into the sql or sql/driver packages and export it so that it can
+// be shared.
+type logicValue string
+
+func (v *logicValue) Scan(value interface{}) error {
+	switch t := value.(type) {
+	case nil:
+		*v = "NULL"
+	case bool:
+		*v = logicValue(strconv.FormatBool(t))
+	case int64:
+		*v = logicValue(strconv.FormatInt(t, 10))
+	case float64:
+		*v = logicValue(strconv.FormatFloat(t, 'g', -1, 64))
+	case []byte:
+		*v = logicValue(t)
+	case string:
+		*v = logicValue(t)
+	case time.Time:
+		*v = logicValue(t.String())
+	default:
+		return fmt.Errorf("unexpected type: %T", value)
+	}
+	return nil
+}
+
+// logicTest executes the test cases specified in a file. The file format is
+// taken from the sqllogictest tool
+// (http://www.sqlite.org/sqllogictest/doc/trunk/about.wiki) with various
+// extensions to allow specifying errors and additional options. See
+// https://github.com/gregrahn/sqllogictest/ for a github mirror of the
+// sqllogictest source.
+//
+// TODO(pmattis): We currently cannot run the tests from sqllogictest due to
+// insufficient SQL coverage (e.g. lack of subqueries and aggregation
+// functions). We should work towards fixing that.
+type logicTest struct {
+	*testing.T
+}
+
+func (t logicTest) run(path string) {
```

Comment:
Once nice effect of keeping the test data in files is that its easy to use the line numbers in the test output. When do we build test binaries and run them separately from the source code?